src/util/newconfig/ConfigConstraints.cpp

```diff
@@ -0,0 +1,197 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/newconfig/ConfigConstraints.hpp"
+
+#include "util/log/Logger.hpp"
+#include "util/newconfig/Error.hpp"
+#include "util/newconfig/Types.hpp"
+
+#include <fmt/core.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <optional>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace util::config {
+
+std::optional<Error>
+PortConstraint::checkTypeImpl(Value const& port) const
```

Comment:
Question about the constraints **PortConstraint**, why don't we use general constraint like _Between_ (0,65535)?
Same Question about **ChannelNameConstraint** and **CassandraName**, more general constraint will be like _OneOf_("info", "debug","warning"...) _OneOf_("cassandra").

Reply:
So for the port constraint, I don't think it can part of `NumberValueConstraint`, as port can be represented by strings in the config, and there are conversions needed from string to int. 

For the OneOf though, I think it can be done! I initially misinterpreted your comment to add "OneOf" constraint as part of ConfigValue, and I thought then that the ConfigValue definition would be wayyy to long, because it will look something like this: 
`{"log_level", ConfigValue{ConfigType::String}.defaultValue("info").withConstraint(OneOf("info", "debug", "warning" ......  ))}.`
but now I recognize that I can hide all the constraint stuff and it can be the same as before. Good catch! 
Thanks @cindyyan317  🙇